submission/proxy_server.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package submission
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/trillian/ctfe"
+)
+
+// ProxyServer wraps Proxy and handles http-requests for it.
+type ProxyServer struct {
+	p          *Proxy
+	addTimeout time.Duration
+}
+
+// NewProxyServer creates and inits ProxyServer instance.
+func NewProxyServer(logListPath string, logListRefreshInterval time.Duration, rootsRefreshInterval time.Duration, dBuilder DistributorBuilder, reqTimeout time.Duration) *ProxyServer {
+	s := &ProxyServer{addTimeout: reqTimeout}
+	s.p = NewProxy(NewLogListRefresher(logListPath), dBuilder)
+	s.p.Run(context.Background(), logListRefreshInterval, rootsRefreshInterval)
+	return s
+}
+
+func marshalSCTs(scts []*AssignedSCT) []byte {
+	var jsonSCTsObj struct {
+		SCTs []ct.SignedCertificateTimestamp `json:"scts"`
+	}
+	jsonSCTsObj.SCTs = make([]ct.SignedCertificateTimestamp, len(scts))
```

Comment:
You probably mean `make([]ct.Blah, 0, len(scts))`. Otherwise, after appends below your slice will have `2*len(scts)`, first half of which is empty.

Reply:
Done